src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -28,6 +30,15 @@ internal sealed partial class JsonClassInfo
 
         public Type Type { get; private set; }
 
+        internal void UpdateSortedPropertyCache(ref ReadStackFrame frame)
+        {
+            // Set the sorted property cache. Overwrite any existing cache which can occur in multi-threaded cases.
+            if (frame.PropertyRefCache != null)
+            {
+                _propertyRefsSorted = frame.PropertyRefCache.ToArray();
```

Comment:
Do we need some kind of synchronization/locking mechanism here?

Reply:
> Do we need some kind of synchronization/locking mechanism here?

That's why the above is volatile.  It avoids out-of-order issues with regards to the data written to the array being published prior to the reference to the array, and similar for the read side.  Technically the volatile isn't necessary due to additional constraints coreclr puts in place, but it is necessary according to the ECMA CLI spec.  And as long as the results here are idempotent, it doesn't matter whether multiple threads race to set the field.